From 9278410f44b4fad6efd2b5d74527394c587eb74e Mon Sep 17 00:00:00 2001
From: Jonathan M Hsieh <jon@cloudera.com>
Date: Wed, 8 Mar 2017 15:51:48 -0800
Subject: [PATCH 307/362] CDH-51354 TestHRegion#testWritesWhileRollWriter
 fails consistently

Revert "HBASE-16429 FSHLog: deadlock if rollWriter called when ring buffer filled with appends"

This reverts commit fe050ceef27f7020b29dfd016fd967ae7c957405.

Change-Id: I53dcc1e380de69296f3e9dc46b28992d97e135aa
Reason: Revert
Author: Jonathan Hsieh
Ref: CDH-51354
---
 .../hadoop/hbase/regionserver/wal/FSHLog.java      |   33 ++-------
 .../hadoop/hbase/regionserver/TestHRegion.java     |   70 --------------------
 2 files changed, 5 insertions(+), 98 deletions(-)

diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/FSHLog.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/FSHLog.java
index da7ed35..0f5e9e6 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/FSHLog.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/FSHLog.java
@@ -831,20 +831,8 @@ public class FSHLog implements WAL {
     // constructor BEFORE the ring buffer is set running so it is null on first time through
     // here; allow for that.
     SyncFuture syncFuture = null;
-    SafePointZigZagLatch zigzagLatch = null;
-    long sequence = -1L;
-    if (this.ringBufferEventHandler != null) {
-      // Get sequence first to avoid dead lock when ring buffer is full
-      // Considering below sequence
-      // 1. replaceWriter is called and zigzagLatch is initialized
-      // 2. ringBufferEventHandler#onEvent is called and arrives at #attainSafePoint(long) then wait
-      // on safePointReleasedLatch
-      // 3. Since ring buffer is full, if we get sequence when publish sync, the replaceWriter
-      // thread will wait for the ring buffer to be consumed, but the only consumer is waiting
-      // replaceWriter thread to release safePointReleasedLatch, which causes a deadlock
-      sequence = getSequenceOnRingBuffer();
-      zigzagLatch = this.ringBufferEventHandler.attainSafePoint();
-    }
+    SafePointZigZagLatch zigzagLatch = (this.ringBufferEventHandler == null)?
+      null: this.ringBufferEventHandler.attainSafePoint();
     afterCreatingZigZagLatch();
     TraceScope scope = Trace.startSpan("FSHFile.replaceWriter");
     try {
@@ -855,11 +843,8 @@ public class FSHLog implements WAL {
       // to come back.  Cleanup this syncFuture down below after we are ready to run again.
       try {
         if (zigzagLatch != null) {
-          // use assert to make sure no change breaks the logic that
-          // sequence and zigzagLatch will be set together
-          assert sequence > 0L : "Failed to get sequence from ring buffer";
           Trace.addTimelineAnnotation("awaiting safepoint");
-          syncFuture = zigzagLatch.waitSafePoint(publishSyncOnRingBuffer(sequence));
+          syncFuture = zigzagLatch.waitSafePoint(publishSyncOnRingBuffer());
         }
       } catch (FailedSyncBeforeLogCloseException e) {
         // If unflushed/unsynced entries on close, it is reason to abort.
@@ -1371,20 +1356,12 @@ public class FSHLog implements WAL {
     return logRollNeeded;
   }
 
-  private SyncFuture publishSyncOnRingBuffer(long sequence) {
-    return publishSyncOnRingBuffer(sequence, null);
-  }
-
-  private long getSequenceOnRingBuffer() {
-    return this.disruptor.getRingBuffer().next();
+  private SyncFuture publishSyncOnRingBuffer() {
+    return publishSyncOnRingBuffer(null);
   }
 
   private SyncFuture publishSyncOnRingBuffer(Span span) {
     long sequence = this.disruptor.getRingBuffer().next();
-    return publishSyncOnRingBuffer(sequence, span);
-  }
-
-  private SyncFuture publishSyncOnRingBuffer(long sequence, Span span) {
     SyncFuture syncFuture = getSyncFuture(sequence, span);
     try {
       RingBufferTruck truck = this.disruptor.getRingBuffer().get(sequence);
diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestHRegion.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestHRegion.java
index 2b0db4c..4d31374 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestHRegion.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestHRegion.java
@@ -6529,74 +6529,4 @@ public class TestHRegion {
     return initHRegion(tableName, callingMethod, HBaseConfiguration.create(),
         families);
   }
-
-  /**
-   * HBASE-16429 Make sure no stuck if roll writer when ring buffer is filled with appends
-   * @throws IOException if IO error occurred during test
-   */
-  @Test(timeout = 60000)
-  public void testWritesWhileRollWriter() throws IOException {
-    int testCount = 10;
-    int numRows = 1024;
-    int numFamilies = 2;
-    int numQualifiers = 2;
-    final byte[][] families = new byte[numFamilies][];
-    for (int i = 0; i < numFamilies; i++) {
-      families[i] = Bytes.toBytes("family" + i);
-    }
-    final byte[][] qualifiers = new byte[numQualifiers][];
-    for (int i = 0; i < numQualifiers; i++) {
-      qualifiers[i] = Bytes.toBytes("qual" + i);
-    }
-
-    String method = "testWritesWhileRollWriter";
-    CONF.setInt("hbase.regionserver.wal.disruptor.event.count", 2);
-    this.region = initHRegion(tableName, method, CONF, families);
-    try {
-      List<Thread> threads = new ArrayList<Thread>();
-      for (int i = 0; i < numRows; i++) {
-        final int count = i;
-        Thread t = new Thread(new Runnable() {
-
-          @Override
-          public void run() {
-            byte[] row = Bytes.toBytes("row" + count);
-            Put put = new Put(row);
-            put.setDurability(Durability.SYNC_WAL);
-            byte[] value = Bytes.toBytes(String.valueOf(count));
-            for (byte[] family : families) {
-              for (byte[] qualifier : qualifiers) {
-                put.addColumn(family, qualifier, (long) count, value);
-              }
-            }
-            try {
-              region.put(put);
-            } catch (IOException e) {
-              throw new RuntimeException(e);
-            }
-          }
-        });
-        threads.add(t);
-      }
-      for (Thread t : threads) {
-        t.start();
-      }
-
-      for (int i = 0; i < testCount; i++) {
-        region.getWAL().rollWriter();
-        Thread.yield();
-      }
-    } finally {
-      try {
-        HBaseTestingUtility.closeRegionAndWAL(this.region);
-        CONF.setInt("hbase.regionserver.wal.disruptor.event.count", 16 * 1024);
-      } catch (DroppedSnapshotException dse) {
-        // We could get this on way out because we interrupt the background flusher and it could
-        // fail anywhere causing a DSE over in the background flusher... only it is not properly
-        // dealt with so could still be memory hanging out when we get to here -- memory we can't
-        // flush because the accounting is 'off' since original DSE.
-      }
-      this.region = null;
-    }
-  }
 }
-- 
1.7.9.5

